File: python3/_110_balanced_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Optional, Union

from python3.helper import tester_helper, Eg, int_list_to_tree
from python3.helper_classes import TreeNode


class Solution:
    def not_bal_or_max_height(self, root: Optional[TreeNode]) -> Union[bool, int]:
        """
        Checks the height of the tree at `root`. If not balanced returns False. Otherwise, returns larger of children
            heights. Note: never returns True.
        :param root: The root of the tree to check the height of
        """
        if root is None:
            return 0
        left_height = self.not_bal_or_max_height(root.left)
        if type(left_height) == bool and not left_height:
            return False
        right_height = self.not_bal_or_max_height(root.right)
        if type(right_height) == bool and not right_height:
            return False
        assert type(left_height) == int and type(right_height) == int
        if abs(left_height - right_height) > 1:
            return False  # Not balanced
        else:
            return max(left_height, right_height) + 1

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        result = self.not_bal_or_max_height(root)
        return type(result) == int


def tester():
    examples = [
        Eg(int_list_to_tree([3, 9, 20, None, None, 15, 7]), True),
        Eg(int_list_to_tree([1, 2, 2, 3, 3, None, None, 4, 4]), False),
        Eg(int_list_to_tree([]), True),
    ]
    tester_helper(110, examples, Solution().isBalanced)

File: rust/src/_110_balanced_binary_tree.rs
//! Solution for https://leetcode.com/problems/balanced-binary-tree
//! 110. Balanced Binary Tree

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

enum BalanceState {
    Balanced { height: usize },
    NotBalanced,
}

impl BalanceState {
    /// Returns `true` if the balance state is [`IsBalanced`].
    ///
    /// [`IsBalanced`]: BalanceState::IsBalanced
    #[must_use]
    fn is_balanced(&self) -> bool {
        matches!(self, Self::Balanced { .. })
    }
}

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::is_balance_and_height(&root).is_balanced()
    }

    fn is_balance_and_height(root: &Option<Rc<RefCell<TreeNode>>>) -> BalanceState {
        let Some(root) = root else {
            return BalanceState::Balanced { height: 0 };
        };
        let root = root.borrow(); // Get borrowed version
        match (
            Self::is_balance_and_height(&root.left),
            Self::is_balance_and_height(&root.right),
        ) {
            (
                BalanceState::Balanced {
                    height: left_height,
                },
                BalanceState::Balanced {
                    height: right_height,
                },
            ) => {
                if left_height.abs_diff(right_height) > 1 {
                    BalanceState::NotBalanced
                } else {
                    BalanceState::Balanced {
                        height: 1 + left_height.max(right_height),
                    }
                }
            }
            _ => BalanceState::NotBalanced,
        }
    }
}

// << ---------------- Code below here is only for local use ---------------- >>

pub struct Solution;
use cargo_leet::TreeNode;

#[cfg(test)]
mod tests {
    use super::*;
    use cargo_leet::TreeRoot;

    use rstest::rstest;

    #[rstest]
    #[case(TreeRoot::from("[3,9,20,null,null,15,7]").into(), true)]
    #[case(TreeRoot::from("[1,2,2,3,3,null,null,4,4]").into(), false)]
    #[case(TreeRoot::from("[]").into(), true)]
    fn case(#[case] root: Option<Rc<RefCell<TreeNode>>>, #[case] expected: bool) {
        let actual = Solution::is_balanced(root);
        assert_eq!(actual, expected);
    }
}
